src/updater/updater.cpp

```diff
@@ -409,27 +435,141 @@ updateTurnPenalties(const UpdaterConfig &config,
+    std::unordered_set<std::tuple<NodeID, NodeID, NodeID>,
+                       std::hash<std::tuple<NodeID, NodeID, NodeID>>>
+        is_no_set;
+    for (const auto &c : conditional_turns)
+    {
+        // only add restrictions to the lookups if the restriction is valid now
+        if (!IsRestrictionValid(time_zone_handler, c, coordinates, osm_node_ids))
+            continue;
+        if (c.flags.is_only)
+        {
+            is_only_lookup.lookup.push_back({std::make_tuple(c.from.node, c.via.node), c.to.node});
+        }
+        else
+        {
+            is_no_set.insert({std::make_tuple(c.from.node, c.via.node, c.to.node)});
+        }
+    }
+
+    for (std::uint64_t edge_index = 0; edge_index < turn_weight_penalties.size(); ++edge_index)
+    {
+        const extractor::lookup::TurnIndexBlock internal_turn = turn_index_blocks[edge_index];
+
+        const auto is_no_tuple =
+            std::make_tuple(internal_turn.from_id, internal_turn.via_id, internal_turn.to_id);
+        const auto is_only_tuple = std::make_tuple(internal_turn.from_id, internal_turn.via_id);
+        // turn has a no_* restriction
+        if (is_no_set.find(is_no_tuple) != is_no_set.end())
+        {
+            util::Log() << "Conditional penalty set on edge: " << edge_index;
+            turn_weight_penalties[edge_index] = INVALID_TURN_PENALTY;
+            updated_turns.push_back(edge_index);
+        }
+        // turn has an only_* restriction
+        else if (is_only_lookup(is_only_tuple))
+        {
+            // with only_* restrictions, the turn on which the restriction is tagged is valid
+            if (*is_only_lookup(is_only_tuple) == internal_turn.to_id)
+                continue;
+
+            util::Log() << "Conditional penalty set on edge: " << edge_index;
+            turn_weight_penalties[edge_index] = INVALID_TURN_PENALTY;
+            updated_turns.push_back(edge_index);
+        }
+    }
+
+    return updated_turns;
+}
 }
 
 Updater::NumNodesAndEdges Updater::LoadAndUpdateEdgeExpandedGraph() const
 {
     std::vector<extractor::EdgeBasedEdge> edge_based_edge_list;
     std::vector<EdgeWeight> node_weights;
-    auto max_edge_id = Updater::LoadAndUpdateEdgeExpandedGraph(edge_based_edge_list, node_weights);
+    std::vector<util::Coordinate> node_coordinates;
+    extractor::PackedOSMIDs osm_node_ids;
+    auto max_edge_id = Updater::LoadAndUpdateEdgeExpandedGraph(
+        edge_based_edge_list, node_weights, node_coordinates, osm_node_ids);
     return std::make_tuple(max_edge_id + 1, std::move(edge_based_edge_list));
 }
 
 EdgeID
 Updater::LoadAndUpdateEdgeExpandedGraph(std::vector<extractor::EdgeBasedEdge> &edge_based_edge_list,
-                                        std::vector<EdgeWeight> &node_weights) const
+                                        std::vector<EdgeWeight> &node_weights,
+                                        std::vector<util::Coordinate> node_coordinates,
```

Comment:
`const std::vector<util::Coordinate>&`, `const extractor::PackedOSMIDs&` otherwise these vectors will be copied.